### PR TITLE
feat: update chat ui to match mockup

### PR DESCRIPTION
**Pull request by nayancprakash**
**Changes**
- change the chat UI to match mockups

**Known Bugs/Differences from mockups to be addressed later**
- bubble.dart does not allow for outlines, this means that the chat bubbles were filled in with colour rather than just outlined. We may need to make a custom bubble class to fix this
- Several logos are different than the mockup, waiting for the logos from @laiiibah 
- The user profile picture was moved to above the user name
- The new match's ano is the incorrect ano, waiting for the correct ano from @laiiibah 
- The bracket is missing from the top right of the chat
- The new matches should have a horizontal fade in/fade out to prevent the harsh clipping against the ano

**UI**
> Screen Shot 2020-07-29 at 3 19 51 PM
> Screen Shot 2020-07-29 at 3 20 00 PM

### CODE REVIEW

File: lib/chat/chat.dart [hunk 1]
```diff
@@ -70,42 +71,13 @@ class _ChatState extends State<Chat> {
...
-}
-
 buildChatView(var newMatchesList, var matchesList) {
   return (
     <Widget>[
-      buildHeader(),
+      GlobalHeader(
+        actionText: "New Message",
+        onActionTap: () => {},
+      ),
```
Comment (eivenlour) [inline lib/chat/chat.dart:80]: I'm assuming you don't need to place this in the constants file for now since it's temporary?

File: lib/chat/chat.dart [hunk 2]
```diff
@@ -70,42 +71,13 @@ class _ChatState extends State<Chat> {
...
       if (newMatchesList.length != 0)
         NewMatches(newMatchesList: newMatchesList),
       if (matchesList.length != 0)
```
Comment (eivenlour) [inline lib/chat/chat.dart:83]: oh wow I didnt know we could do if statements without the brackets!

File: lib/chat/chat_messenger.dart [hunk 3]
```diff
@@ -157,49 +158,63 @@ class ChatMessengerState extends State<ChatMessenger> {
     return Container(
```
Comment (eivenlour) [inline lib/chat/chat_messenger.dart:158]: This whole return statement is kinda overwhelming, so it's probably better to break it into multiple widget functions?

File: lib/chat/chat_messenger.dart [hunk 4]
```diff
@@ -274,25 +289,48 @@ class ChatMessengerState extends State<ChatMessenger> {
...
+                  )
+                ),
+                Text(
+                  widget.recipient,
+                  style: GoogleFonts.muli(
+                    textStyle: TextStyle(
+                      color: Colors.black, 
+                      letterSpacing: .5, 
+                      fontSize: 14.0, 
+                      fontWeight: FontWeight.bold
```
Comment (eivenlour) [inline lib/chat/chat_messenger.dart:328]: You can use `FormatText` here from the common folder!

File: lib/chat/chat_messenger.dart [hunk 5]
```diff
@@ -274,25 +289,48 @@ class ChatMessengerState extends State<ChatMessenger> {
...
-                  fontWeight: FontWeight.bold
-                )
+              icon: Icon(
+                Icons.arrow_back_ios,
+                color: Theme.of(context).accentColor
               )
             ),
+            backgroundColor: Colors.white,
+            centerTitle: true,
+            title: Column (
```
Comment (eivenlour) [inline lib/chat/chat_messenger.dart:305]: Same here! Maybe make a separate function that returns the title widget so it's not overwhelming

File: lib/chat/message.dart [hunk 6]
```diff
@@ -16,50 +17,54 @@ class MessageView extends StatelessWidget {
...
+          child: Text(
+            message,
+            style: GoogleFonts.muli(
+              textStyle: TextStyle(
+                color: Colors.black, 
+                letterSpacing: .5, 
+                fontSize: 14.0, 
+              )
             )
           )
```
Comment (eivenlour) [inline lib/chat/message.dart:40]: Make a `buildTextMessage()` function?

File: lib/chat/message.dart [hunk 7]
```diff
@@ -16,50 +17,54 @@ class MessageView extends StatelessWidget {
...
+              placeholder: (context, url) => Container(
+                width: 200.0,
+                height: 200.0,
+                padding: EdgeInsets.all(70.0),
+                child: CircularProgressIndicator(
+                  valueColor: AlwaysStoppedAnimation<Color>(Colors.white)
+                )
+              ),
+              errorWidget: (context, url, error) => Text("error"),
+            )
```
Comment (eivenlour) [inline lib/chat/message.dart:65]: Same here! make a different function and call it here

File: lib/chat/user_message_rows.dart [hunk 8]
```diff
@@ -90,65 +77,100 @@ buildMessageRows(
...
-              width: 80.0,
-              height: 80.0,
-              child: Center(
-                child: Container(
-                  width: 60.0, 
-                  height: 60.0, 
-                  decoration: BoxDecoration(
-                    color: Colors.black, 
-                    shape: BoxShape.circle
+        child: Container(
```
Comment (eivenlour) [inline lib/chat/user_message_rows.dart:80]: Same with this whole child!

File: lib/chat/chat_messenger.dart [hunk 9]
```diff
@@ -157,49 +158,63 @@ class ChatMessengerState extends State<ChatMessenger> {
...
-              onSubmitted: (value) => submitMessage(textMessage, value),
+              child: Row(
+                crossAxisAlignment: CrossAxisAlignment.end,
+                children: <Widget>[
+                  IconButton(
+                    icon: Icon(Icons.photo),
+                    iconSize: 24.0,
+                    padding: EdgeInsets.all(0.0),
+                    color: Theme.of(context).accentColor,
+                    onPressed: () => pickImage(ImageSource.gallery),
```
Comment (eivenlour) [inline lib/chat/chat_messenger.dart:181]: Also, you can probs add this `IconButton` widget in `styles.dart` in the common folder since you've called it more than once and just pass `iconData` and `onPressed` arguments

File: lib/chat/chat.dart [hunk 10]
```diff
@@ -4,6 +4,7 @@ import 'package:flutter/material.dart';
 import 'package:flutter_redux/flutter_redux.dart';
```
Comment (HelenGezahegn) [inline lib/chat/chat.dart:4]: You can delete the import statement from `google_fonts`.

File: lib/chat/chat.dart (hunk 2 above)
Comment (HelenGezahegn) [reply lib/chat/chat.dart:83]: If it's a one-liner thing, yes! Although I know in React, linters complains about it.

File: lib/chat/chat_messenger.dart [hunk 11]
```diff
@@ -157,49 +158,63 @@ class ChatMessengerState extends State<ChatMessenger> {
...
+                    iconSize: 24.0,
+                    padding: EdgeInsets.all(0.0),
+                    color: Theme.of(context).accentColor,
+                    onPressed: () => pickImage(ImageSource.camera),
+                  ),
+                  Expanded(
+                    child: TextField(
+                      textCapitalization: TextCapitalization.sentences,
+                      decoration: InputDecoration.collapsed(
+                        hintText: 'Send a message...',
```
Comment (HelenGezahegn) [inline lib/chat/chat_messenger.dart:194]: This can probably be put in a constants file.

File: lib/chat/chat_messenger.dart [hunk 12]
```diff
@@ -157,49 +158,63 @@ class ChatMessengerState extends State<ChatMessenger> {
...
+                      ),
+                      style: GoogleFonts.muli(
```
Comment (HelenGezahegn) [inline lib/chat/chat_messenger.dart:196]: Can use the `FormatText` widget here.

File: lib/chat/user_message_rows.dart [hunk 13]
```diff
@@ -73,6 +59,7 @@ buildMessageRows(
     var lastMessage = match.lastMessage;
     var lastMessageType = lastMessage.type;
     var isSent = lastMessage.idFrom == id;
+    var isRead = !isSent && lastMessage.isRead == true;
```
Comment (HelenGezahegn) [inline lib/chat/user_message_rows.dart:62]: Do you need the `== true` part? I think it's automatically interpreted as a boolean with the `&&` operator.

File: lib/chat/user_message_rows.dart [hunk 14]
```diff
@@ -90,65 +77,100 @@ buildMessageRows(
...
                         )
-                      ]
+                      ],
                     ),
-                  )
-                ]
+                    RichText(
+                      overflow: TextOverflow.ellipsis,
+                      text: TextSpan(
+                        style: GoogleFonts.muli(
```
Comment (HelenGezahegn) [inline lib/chat/user_message_rows.dart:138]: `FormatText` widget can be used here too.

File: lib/chat/user_message_rows.dart [hunk 15]
```diff
@@ -90,65 +77,100 @@ buildMessageRows(
...
+                          textStyle: TextStyle(
+                            color: Colors.grey,
+                            letterSpacing: .5, 
+                            fontSize: 12.0, 
+                          )
+                        ),
+                        children: [
+                          if(isSent != false) WidgetSpan(
```
Comment (HelenGezahegn) [inline lib/chat/user_message_rows.dart:146]: I think since this is a multi-line widget below the if, it'd be a good idea to include brackets to wrap around the widgets controlled by the condition.

File: lib/chat/message.dart [hunk 16]
```diff
@@ -16,50 +17,54 @@ class MessageView extends StatelessWidget {
...
+          style: BubbleStyle(
+            margin: isSent ? 
+              BubbleEdges.fromLTRB(50.0, 10.0, 10.0, 10.0) :
+              BubbleEdges.fromLTRB(10.0, 10.0, 50.0, 10.0),
+            padding: BubbleEdges.all(10.0),
+            color: isSent ? Theme.of(context).accentColor : Colors.grey,
+          ),
+          child: Text(
+            message,
+            style: GoogleFonts.muli(
```
Comment (HelenGezahegn) [inline lib/chat/message.dart:33]: Can use the `FormatText` here.

File: lib/chat/message.dart [hunk 17]
```diff
@@ -16,50 +17,54 @@ class MessageView extends StatelessWidget {
...
-              child: CircularProgressIndicator(
-                valueColor: AlwaysStoppedAnimation<Color>(Colors.white)
-              )
-            ),
-            errorWidget: (context, url, error) => Text("error"),
+        ) : Bubble(
+          nip: isSent ? BubbleNip.rightTop : BubbleNip.leftTop,
+          nipOffset: 0.0,
+          style: BubbleStyle(
+            margin: isSent ?
```
Comment (HelenGezahegn) [inline lib/chat/message.dart:45]: Since `isSent` is being used multiple times in this file, this can be checked at the beginning of the widget to determine which styling to use.

**Comment by HelenGezahegn** [comment]
Also, how do I test this PR? I pulled it locally and the chat screen is empty. I'm assuming it's because I have no dummy data?

File: lib/chat/chat_messenger.dart (hunk 12 above)
Comment (nayancprakash) [reply lib/chat/chat_messenger.dart:196]: don't think I can, this is a `TextField` whereas `FormatText` is just a `Text`

File: lib/chat/user_message_rows.dart (hunk 14 above)
Comment (nayancprakash) [reply lib/chat/user_message_rows.dart:138]: this needs to be a `TextSpan` whereas the `FormatText` is a `Text` because this needs to be able to be inserted as part of the `RichText`

File: lib/chat/user_message_rows.dart (hunk 15 above)
Comment (nayancprakash) [reply lib/chat/user_message_rows.dart]: I think the brackets around the widget actually returns them as a Set. I'll double check this again when its not 2:30 in the morn

File: lib/chat/chat_messenger.dart (hunk 12 above)
Comment (eivenlour) [reply lib/chat/chat_messenger.dart]: I think aside from `FormatText`, we should also make a general `FormatTextStyle` widget since I've also been needing it lately for my form fields  👀 So ya, you can leave this be for now! @nayancprakash

**Comment by nayancprakash** [comment]
> Also, how do I test this PR? I pulled it locally and the chat screen is empty. I'm assuming it's because I have no dummy data?

you may have to manually mock a match on Firestore, look at some of the other matches as an example
